src/EFCore.PG/Scaffolding/Internal/NpgsqlDatabaseModelFactory.cs

```diff
@@ -511,191 +565,231 @@ NOT indisprimary
 
-                        var principalColumns = (List<DatabaseColumn>)principalTable.Columns;
+                            if (columnIndices.Length != principalColumnIndices.Length)
+                                throw new InvalidOperationException("Found varying lengths for column and principal column indices.");
 
-                        for (var i = 0; i < columnIndices.Length; i++)
-                        {
-                            var foreignKeyColumn = table.Columns[columnIndices[i] - 1];
-                            var foreignKeyPrincipalColumn = principalColumns[principalColumnIndices[i] - 1];
-                            if (foreignKeyColumn == null || foreignKeyPrincipalColumn == null)
+                            var principalColumns = (List<DatabaseColumn>)principalTable.Columns;
+
+                            for (var i = 0; i < columnIndices.Length; i++)
                             {
-                                _logger.UnsupportedColumnConstraintSkippedWarning(foreignKey.Name, DisplayName(tableSchema, tableName));
-                                goto ForeignKeyEnd;
+                                var foreignKeyColumn = table.Columns[columnIndices[i] - 1];
+                                var foreignKeyPrincipalColumn = principalColumns[principalColumnIndices[i] - 1];
+                                if (foreignKeyColumn == null || foreignKeyPrincipalColumn == null)
+                                {
+                                    _logger.UnsupportedColumnConstraintSkippedWarning(foreignKey.Name, DisplayName(tableSchema, tableName));
+                                    goto ForeignKeyEnd;
+                                }
+
+                                foreignKey.Columns.Add(foreignKeyColumn);
+                                foreignKey.PrincipalColumns.Add(foreignKeyPrincipalColumn);
                             }
-                            foreignKey.Columns.Add(foreignKeyColumn);
-                            foreignKey.PrincipalColumns.Add(foreignKeyPrincipalColumn);
-                        }
 
-                        table.ForeignKeys.Add(foreignKey);
-                    ForeignKeyEnd: ;
-                    }
+                            table.ForeignKeys.Add(foreignKey);
+                            ForeignKeyEnd: ;
+                        }
 
-                    // Unique constraints
-                    foreach (var record in tableGroup
-                        .Where(ddr => ddr.GetValueOrDefault<char>("contype") == 'u')
-                        .ToArray())
-                    {
-                        var name = record.GetValueOrDefault<string>("conname");
+                        // Unique constraints
+                        foreach (var record in tableGroup
+                                               .Where(ddr => ddr.GetValueOrDefault<char>("contype") == 'u')
+                                               .ToArray())
```

Comment:
Decrease indentation. You know that the methods are related to the collection. But there is no reason to use `ToArray`, so you can move filtration to the same line where you iterates over `tableGroups`.

Reply:
I wasn't sure if this might be here to guard against a collection modification during enumeration. But having looked at everything else, you're probably right that it isn't needed. Removing.